### PR TITLE
Updated parameters of the Mustache SC algorithm, optimized for Run 3

### PR DESCRIPTION
**Pull request by rchatter**
This PR updates the hard coded parameters defining the Mustache superclustering algorithm used to cluster electrons and photons in the CMS ECAL. Here the optimization has been performed based on expected Run 3 conditions (but doesn't break anything if used with Run 2 conditions).

 A more elegant implementation is underway where these parameters would be loaded from the DB, but this will not meet the last 11_2_X pre-release deadline next Wednesday, hence the hardcoded parameters are being updated.

These parameters have been approved by the ECAL DPG. The details of derivation and validation of the same may be seen in these slides from N. Marinelli and  L. Zygala https://indico.cern.ch/event/949294/contributions/3988389/attachments/2091573/3514649/2020_08_26_Clustering.pdf

 @nancymarinelli @lzygala @thomreis @fcouderc  @simonepigazzini @mdonega

### CODE REVIEW

**Comment by jpata** [comment]
@cmsbuild please test

File: RecoEcal/EgammaCoreTools/src/Mustache.cc [hunk 1]
```diff
@@ -68,25 +68,25 @@ namespace reco {
     bool inDynamicDPhiWindow(
         const float seedEta, const float seedPhi, const float ClustE, const float ClusEta, const float ClusPhi) {
       // from Rishi's fits 06 June 2013 in log base 10
```
Comment (perrotta) [inline RecoEcal/EgammaCoreTools/src/Mustache.cc:70]: The new parameters are not taken from the Rishi's fit  of more than 7 years ago, I imagine: either you update the comment, or better to remove it

Comment (rchatter) [reply RecoEcal/EgammaCoreTools/src/Mustache.cc]: Updated the comment to add the correct reference.
Thanks,
Rajdeep

**Comment by perrotta** [comment]
please test

**Comment by jpata** [comment]
@rchatter we ran some additional validation on this PR with gun samples to be sure:

We see an increase of photon multiplicity per jet (16.0):
[DQMData__Run1__JetMET__Runsummary__JetValidation__slimmedJets__photonMultiplicity.pdf](https://github.com/cms-sw/cmssw/files/5438465/DQMData__Run1__JetMET__Runsummary__JetValidation__slimmedJets__photonMultiplicity.pdf)

We also see a slightly better jet resolution (19.0):
[DQMData__Run1__JetMET__Runsummary__JetValidation__ak4PFJetsCHS__PtRecoOverGen_E_20_40.pdf](https://github.com/cms-sw/cmssw/files/5438474/DQMData__Run1__JetMET__Runsummary__JetValidation__ak4PFJetsCHS__PtRecoOverGen_E_20_40.pdf)

Here's a list of all the DQM plots with MET or JetValidation:
http://jpata.web.cern.ch/jpata/reco/31823/16.0/
http://jpata.web.cern.ch/jpata/reco/31823/19.0/

Let me know if this is in line with your expectations and if you see anything else out of the ordinary in the DQM plots. Thanks!

**Comment by jpata** [comment]
Here's phase2 upgrade workflow DQM, please check this one, instead of the Run1 I posted above (thx Slava for noticing!)

DQMData__Run1__JetMET__Runsummary__JetValidation__ak4PFJetsCHS__photonMultiplicity

http://jpata.web.cern.ch/jpata/reco/31823/11705.0/
http://jpata.web.cern.ch/jpata/reco/31823/11602.0/

**Comment by simonepigazzini** [comment]
Hi @jpata, do you have DQM plots for ECAL superclusters or pfPhotons/Electrons with a e/gamma gun sample? The mustache is really only used in the reco of e/gamma so any effect on jets in an indirect one. My guess here is that having a tighter parameters in the mustache superclustering means that we cluster less energy from PU into superclusters and therefore the jets peaks up more photon.

Thank you,

simone

**Comment by simonepigazzini** [comment]
Hi again,

there should be a Hgg sample run as part of the standard validation matrix, @nancymarinelli could have a look at that if you saved the DQM from that workflow. 

Here https://lzygala.web.cern.ch/lzygala/Envelope_OptPFRecHitThresholds34/mustache_comparison_SingleSetParams_Weighted/h_SC_nPFClusters.pdf you can also see the number of pfClusters (i.e. the same thing as photon multiplicity but for ECAL SC) with the old and new set of parameters. For ECAL SC the number of constituents decreases with the new parameters so it can be reasonable to expect that the rejected pfClusters are clustered by nearby jets, this would explain the plot above.

Could you also please confirm that these are Run2 DQM plots and which release is "ref"?

Thank you,

simone

**Comment by jpata** [comment]
Thx for the suggestion @simonepigazzini! The idea here was to check downstream things that might not otherwise have been in the DPG view in the validations, but we can also produce plots for specific EGamma validation. For the photon gun sample, here are a few that see bigger changes with this update. The rest of the plots are at the cern link above.

DQMData__Run1__EgammaV__Runsummary__pfPhotonValidator__Efficiencies__convEffVsEtTotal
DQMData__Run1__Egamma__Runsummary__gedPhotonAnalyzer__AllPhotons__Etabove20GeV__h_15_e1x5VsEt


> Could you also please confirm that these are Run2 DQM plots and which release is "ref"?

11602.0 and 11705.0 are run3 workflows as far as I understand (2021), ref was `CMSSW_11_2_X_2020-10-22-1100` without this PR included.

From the reco point of view, I would say as long as the changes are understood and noticed by the relevant POGs also downstream, this is acceptable and we can proceed. :)

**Comment by jpata** [comment]
Just to be sure of no surprises downstream, I ran some extended tests. While I don't find any major differences in downstream jet or met outputs, there are some fluctuations.

For example, `11630.0_QCD_Pt_3000_3500_14TeV+2021` on `CMSSW_11_2_X_2020-11-01-2300`:
DQMData__Run1__JetMET__Runsummary__METValidation__pfMet__METResolution_GenMETTrue_InMETBins
DQMData__Run1__JetMET__Runsummary__JetValidation__slimmedJets__photonMultiplicity

The full DQM output for `JetValidation\|METValidation` can be found here:
http://jpata.web.cern.ch/jpata/reco/31823/11630.0/
http://jpata.web.cern.ch/jpata/reco/31823/11671.0/
http://jpata.web.cern.ch/jpata/reco/31823/11705.0/

Overall, I think we can proceed with this PR from the reco side.

**Comment by rchatter** [comment]
Hi Joosep,

Just to understand the second plot: Is the comparison on precisely the same events with and w/o this update, in which case the difference seems a "little" large. Also I precisely remember how exactly the photon multiplicity in a jet is evaluated. But most likely these are not photons using the Mustache SC algorithm so like you said shouldn't be a cause for worry [unless JETMET thinks otherwise].

Best,
Rajdeep

**Comment by jpata** [comment]
These should be the same events, generated with the workflows mentioned above. It is my understanding that running the GEN-SIM workflow twice with the same seed should give exactly the same outputs, modulo changes brought in by this PR.

**Comment by simonepigazzini** [comment]
Hi All,

I think that despite the events being the same a change in what goes into the superclusters means that we can end up looking at different jets. I would say that there are no major trend differences w/ and w/o this PR. On the ECAL side we are confident that the new parameters are ok for the superclusters and we will soon have a deeper look at what happens with jets (validation with more events). I think the PR can be merged, keeping in mind that soon we will have a new PR moving these parameters to the database, at which point we will be easily able to fix any problem with the parameters on the fly through a new GT.

Best,

simone

**Comment by jpata** [comment]
+reconstruction

- updates ECAL Moustache supercluster parameters for Run3 directly in the algorithm implementation
- jenkins tests and DQM result in a large number of differences for EGamma-related outputs, which are understood
- EGamma performance has been validated in the ECAL DPG
- based on quick tests on the reco side, we don't observe major trends in downstream variables such as jets or MET

**Comment by qliphy** [comment]
+1

**Pull request merged**